src/storage/src/table/state_table.rs

```diff
@@ -102,28 +111,128 @@ impl<S: StateStore> StateTable<S> {
         Ok(())
     }
 
-    pub async fn iter(&self, _pk: Row) -> StorageResult<StateTableRowIter<S>> {
-        todo!()
+    pub async fn iter(&self, epoch: u64) -> StorageResult<StateTableRowIter<'_, S>> {
+        let mem_table_iter = self.mem_table.buffer.iter().peekable();
+        StateTableRowIter::new(
+            &self.keyspace,
+            &self.column_descs,
+            mem_table_iter,
+            &self.order_types,
+            epoch,
+        )
+        .await
     }
 }
 
-pub struct StateTableRowIter<S: StateStore> {
-    cell_based_iter: CellBasedTableRowIter<S>,
-    mem_table: MemTable,
+/// `StateTableRowIter` is able to read the just written data (uncommited data).
+/// It will merge the result of `mem_table_iter` and `cell_based_streaming_iter`
+pub struct StateTableRowIter<'a, S: StateStore> {
+    mem_table_iter: Peekable<MemTableIter<'a>>,
+    cell_based_streaming_iter: CellBasedTableStreamingIter<S>,
+    /// The result of the last cell_based_streaming_iter next is saved, which can avoid being
+    /// discarded.
+    pk_serializer: OrderedRowSerializer,
 }
+type MemTableIter<'a> = Iter<'a, Row, RowOp>;
```

Comment:
Better to qualify it more detailedly.
```suggestion
type MemTableIter<'a> = btree_map::Iter<'a, Row, RowOp>;
```